### ISSUE TITLE
Cythonize Entropy

### ISSUE THREAD

**Issue by Eve-ning** [Eve-ning/FRModel#43]

**Comment by Eve-ning** [comment]
Initial Draft
```
@cython.boundscheck(False)
@cython.wraparound(False)
def entropy_256_3dim_w(np.ndarray[DTYPE_t, ndim=3] a,
                       np.ndarray[DTYPE_t, ndim=3] b,
                       unsigned int radius):
    """ Entropy Calculation in Cython

    Limitations:
    Array Values cannot exceed 255, excluding 255.
    Array Values must be 3 dim, last dim will be iterated through
    Entropy result will overflow on > unsigned int
    Maximum GLCM occurrence cannot exceed 65535

    Uses entropy_256_2dim for looping last dimension

    """

    """
    O O O           X X O   Window Size = 2
    O O O if 2x2 -> X X O : Window I Rows = 2
    O O O           O O O   Window I Cols = 2
    
    Window I Rows defines how many possible windows can be slid on the frame vertically.
    Same with Columns
    
    Within each Window, we define 
    
    X X  : Window Rows = 2
    X X    Window Cols = 2
    
    This is the explicit number of rows and columns the window has
    """

    assert a.shape[0] == b.shape[0]
    assert a.shape[1] == b.shape[1]
    assert a.shape[2] == b.shape[2]


    cdef DTYPE_t [:, :, :] a_view = a
    cdef DTYPE_t [:, :, :] b_view = b

    cdef unsigned int w_size = radius * 2 + 1
    cdef unsigned int wi_rows = (<unsigned int> a.shape[0]) - w_size + 1
    cdef unsigned int wi_cols = (<unsigned int> a.shape[1]) - w_size + 1
    cdef size_t wi_r = 0, wi_c = 0

    cdef unsigned int w_rows = (<unsigned int> a.shape[0]) - w_size - 1
    cdef unsigned int w_cols = (<unsigned int> a.shape[1]) - w_size - 1
    cdef size_t w_r = 0, w_c = 0

    cdef unsigned char w_channels = <char>a.shape[2]
    cdef size_t w_ch = 0

    entropy_ar = np.zeros([wi_rows, wi_cols, w_channels], dtype=np.uint32)
    cdef unsigned int[:, :, :] entropy_view = entropy_ar

    cdef unsigned short glcm_size = 256
    cdef unsigned short glcm[256][256]
    cdef unsigned short [:, :] glcm_view = glcm
    cdef size_t glcm_r = 0, glcm_c = 0

    cdef unsigned int entropy = 0
    cdef unsigned char entropy_power = 2

    for wi_r in range(wi_rows):
        for wi_c in range(wi_cols):
            # Slide through possible window top lefts

            # Reset GLCM View
            glcm_view[:, :] = 0

            for w_ch in range(w_channels):
                for w_r in range(w_size):
                    for w_c in range(w_size):
                        glcm_view[a_view[wi_r + w_r, wi_c + w_c, w_ch],
                                  b_view[wi_r + w_r, wi_c + w_c, w_ch]] += 1 # Make into const?
                entropy = 0
                for glcm_r in range(glcm_size):
                    for glcm_c in range(glcm_size):
                        entropy += glcm_view[glcm_r, glcm_c] ** entropy_power
                entropy_view[wi_r, wi_c, w_ch] = entropy

    return entropy_ar
```

approx 5.5 ~ 6.0s

**Comment by Eve-ning** [comment]
Second Draft
uses a summed 256 offset  method

```

ctypedef np.npy_uint16 DTYPE_t2

@cython.boundscheck(False)
@cython.wraparound(False)
def entropy(np.ndarray[DTYPE_t2, ndim=3] c,
            unsigned int radius):
    """ Entropy Calculation in Cython

    Limitations:
    Array Values cannot exceed 255, excluding 255.
    Array Values must be 3 dim, last dim will be iterated through
    Entropy result will overflow on > unsigned int
    Maximum GLCM occurrence cannot exceed 65535

    Uses entropy_256_2dim for looping last dimension

    """

    """
    O O O           X X O   Window Size = 2
    O O O if 2x2 -> X X O : Window I Rows = 2
    O O O           O O O   Window I Cols = 2
    
    Window I Rows defines how many possible windows can be slid on the frame vertically.
    Same with Columns
    
    Within each Window, we define 
    
    X X  : Window Rows = 2
    X X    Window Cols = 2
    
    This is the explicit number of rows and columns the window has
    """

    cdef DTYPE_t2 [:, :, :] c_view = c

    cdef unsigned int w_size = radius * 2 + 1
    cdef unsigned int wi_rows = (<unsigned int> c.shape[0]) - w_size + 1
    cdef unsigned int wi_cols = (<unsigned int> c.shape[1]) - w_size + 1
    cdef size_t wi_r = 0, wi_c = 0

    cdef unsigned int w_rows = (<unsigned int> c.shape[0]) - w_size - 1
    cdef unsigned int w_cols = (<unsigned int> c.shape[1]) - w_size - 1
    cdef size_t w_r = 0, w_c = 0

    cdef unsigned char w_channels = <char>c.shape[2]
    cdef size_t w_ch = 0

    entropy_ar = np.zeros([wi_rows, wi_cols, w_channels], dtype=np.uint32)
    cdef unsigned int[:, :, :] entropy_view = entropy_ar

    cdef unsigned short glcm_size = 65536
    cdef unsigned short glcm[65536]
    cdef unsigned short [:] glcm_view = glcm
    cdef size_t glcm_r = 0, glcm_c = 0

    cdef unsigned int entropy = 0
    cdef unsigned char entropy_power = 2

    for wi_r in range(wi_rows):
        for wi_c in range(wi_cols):
            # Slide through possible window top lefts

            # Reset GLCM View
            glcm_view[:] = 0

            for w_ch in range(w_channels):
                for w_r in range(w_size):
                    for w_c in range(w_size):
                        glcm_view[c_view[wi_r + w_r, wi_c + w_c, w_ch]] += 1 # Make into const?
                entropy = 0
                for glcm_i in range(glcm_size):
                    entropy += glcm_view[glcm_i] ** entropy_power
                entropy_view[wi_r, wi_c, w_ch] = entropy

    return entropy_ar
```

approx 1.0s

**Comment by Eve-ning** [comment]
Final Draft

```
import numpy as np
cimport numpy as np
cimport cython
ctypedef np.npy_uint16 DTYPE_t
ctypedef np.npy_uint32 DTYPE_t32

@cython.boundscheck(False)
@cython.wraparound(False)
def cy_entropy(np.ndarray[DTYPE_t, ndim=3] c,
              unsigned int radius):
    """ Entropy Calculation in Cython

    Requirements:
    Array C is defined as 256 * A + B.
    Array C Must be of np.uint16.
    This is to get a 1-to-1 unique tuple-like mapping.

    Limitations:
    Array Values cannot exceed 255, excluding 255.
    Array Values must be 3 dim, last dim will be iterated through
    Entropy result will overflow on > unsigned int
    Maximum GLCM occurrence cannot exceed 65535

    Uses entropy_256_2dim for looping last dimension

    """


    # O O O           X X O   Window Size = 2
    # O O O if 2x2 -> X X O : Window I Rows = 2
    # O O O           O O O   Window I Cols = 2
    #
    # Window I Rows defines how many possible windows can be slid on the frame vertically.
    # Same with Columns
    #
    # Within each Window, we define
    #
    # X X  : Window Rows = 2
    # X X    Window Cols = 2
    #
    # This is the explicit number of rows and columns the window has

    cdef DTYPE_t [:, :, :] c_view = c

    cdef unsigned short w_size = radius * 2 + 1
    cdef unsigned short wi_rows = (<unsigned int> c.shape[0]) - w_size + 1
    cdef unsigned short wi_cols = (<unsigned int> c.shape[1]) - w_size + 1
    cdef size_t wi_r = 0, wi_c = 0

    cdef unsigned int w_rows = (<unsigned int> c.shape[0]) - w_size - 1
    cdef unsigned int w_cols = (<unsigned int> c.shape[1]) - w_size - 1
    cdef size_t w_r = 0, w_c = 0

    cdef unsigned char w_channels = <char>c.shape[2]
    cdef size_t w_ch = 0

    cdef np.ndarray[DTYPE_t32, ndim=3] entropy_ar = np.zeros([wi_rows, wi_cols, w_channels], dtype=np.uint32)
    cdef unsigned int[:, :, :] entropy_view = entropy_ar

    cdef unsigned short glcm_size = <unsigned short>65536
    cdef unsigned short glcm[65536]
    cdef unsigned short [:] glcm_view = glcm
    cdef size_t glcm_i = 0

    cdef unsigned int entropy = 0
    cdef unsigned char entropy_power = 2

    # For each valid window row and column:
    #   We loop through channels, then rows, then columns of each window
    #       In each window we throw the value into a GLCM 65535 long array
    #       In that window, we then loop through the array and square sum
    #       We slot this value into the corresponding cell in entropy_ar
    for wi_r in range(wi_rows):
        for wi_c in range(wi_cols):
            # Slide through possible window top lefts
            glcm_view[:] = 0

            for w_ch in range(w_channels):
                for w_r in range(w_size):
                    for w_c in range(w_size):
                        glcm_view[c_view[wi_r + w_r, wi_c + w_c, w_ch]] += 1
                entropy = 0

                for glcm_i in range(glcm_size):
                    entropy += glcm_view[glcm_i] ** entropy_power
                entropy_view[wi_r, wi_c, w_ch] = entropy

    return entropy_ar
```
approx 1.1s ~ 1.4s

**Comment by Eve-ning** [comment]
Fixed some issues pertaining to `ushort` overflowing to 0 in `glcm_size`

```
import numpy as np
cimport numpy as np
cimport cython
ctypedef np.uint16_t DTYPE_t16
ctypedef np.uint32_t DTYPE_t32
from cython.parallel cimport prange
from tqdm import tqdm

@cython.boundscheck(False)
@cython.wraparound(False)
def cy_entropy(np.ndarray[DTYPE_t16, ndim=3] c,
               unsigned int radius,
               bint verbose):
    """ Entropy Calculation in Cython

    Requirements:
    Array C is defined as 256 * A + B.
    Array C Must be of np.uint16.
    This is to get a 1-to-1 unique tuple-like mapping.

    Limitations:
    Array Values cannot exceed 255, excluding 255.
    Array Values must be 3 dim, last dim will be iterated through
    Entropy result will overflow on > unsigned int
    Maximum GLCM occurrence cannot exceed 65535

    Verbose, whether to output progress with a progress bar

    Uses entropy_256_2dim for looping last dimension

    """

    # O O O           X X O   Window Size = 2
    # O O O if 2x2 -> X X O : Window I Rows = 2
    # O O O           O O O   Window I Cols = 2
    #
    # Window I Rows defines how many possible windows can be slid on the frame vertically.
    # Same with Columns
    #
    # Within each Window, we define
    #
    # X X  : Window Rows = 2
    # X X    Window Cols = 2
    #
    # This is the explicit number of rows and columns the window has

    cdef DTYPE_t16 [:, :, :] c_view = c

    cdef unsigned short w_size = radius * 2 + 1
    cdef unsigned short wi_rows = (<unsigned int> c.shape[0]) - w_size + 1
    cdef unsigned short wi_cols = (<unsigned int> c.shape[1]) - w_size + 1
    cdef short wi_r = 0, wi_c = 0

    cdef unsigned int w_rows = (<unsigned int> c.shape[0]) - w_size - 1
    cdef unsigned int w_cols = (<unsigned int> c.shape[1]) - w_size - 1
    cdef short w_r = 0, w_c = 0

    cdef unsigned char w_channels = <char>c.shape[2]
    cdef char w_ch = 0

    entropy_ar = np.zeros([wi_rows, wi_cols, w_channels], dtype=np.uint32)
    cdef np.uint32_t[:, :, :] entropy_view = entropy_ar
    entropy_view[:, :, :] = 0

    cdef int glcm_size = <unsigned int>65536
    cdef unsigned short glcm[65536]
    cdef unsigned short [:] glcm_view = glcm
    cdef int glcm_i = 0

    cdef unsigned int entropy = 0
    cdef unsigned char entropy_power = 2

    # For each valid window row and column:
    #   We loop through channels, then rows, then columns of each window
    #       In each window we throw the value into a GLCM 65535 long array
    #       In that window, we then loop through the array and square sum
    #       We slot this value into the corresponding cell in entropy_ar

    # The outer loop is required for verbose tqdm
    # prange will not work if there are any Python objects within
    # So the options are exclusive, but verbose is more important
    for wi_r in tqdm(range(wi_rows), disable=not verbose, desc="Entropy Progress"):
        for wi_c in prange(wi_cols, nogil=True, schedule='dynamic'):
            # Slide through possible window top lefts
            glcm_view[:] = 0

            for w_ch in prange(w_channels, schedule='dynamic'):
                for w_r in prange(w_size, schedule='dynamic'):
                    for w_c in prange(w_size, schedule='dynamic'):
                        glcm_view[<int>c_view[wi_r + w_r, wi_c + w_c, w_ch]] += 1

                entropy = 0
                for glcm_i in prange(glcm_size, schedule='dynamic'):
                    entropy += glcm_view[glcm_i] ** entropy_power
                entropy_view[wi_r, wi_c, w_ch] += entropy

    return entropy_ar

```
Approx 4s

I tested with a larger image, it seems to be miles better when the images are larger, however, slower than the current `numpy` method if it's a small image.

This is likely due to overestimating the **GLCM** occurences. However, we'll leave it as is because we're expecting larger images in the research

**Comment by Eve-ning** [comment]
We could do the following within each window also

E.g.
```
ABAACED 
> SORT
AAABCDE
> LOOP
>> A - A - A [3 ** 2]
>> B [1 ** 2]
>> C [1 ** 2]
>> D [1 ** 2]
>> E [1 ** 2]
> SUM
13 
```

Issue is with the memory view being annoying to flatten, not sure how to do that

**Comment by Eve-ning** [comment]
Fixed a major issue in #56, where ``glcm_view`` was incorrectly placed take note
